resolver/src/p2p/peer/actor.rs

```diff
@@ -0,0 +1,303 @@
+    Key: Array,
+    Con: Consumer<Key = Key, Value = Value, FailureCode = ()>,
+    Pro: Producer<Key = Key, Value = Value>,
+    NetS: Sender<PublicKey = C::PublicKey>,
+    NetR: Receiver<PublicKey = C::PublicKey>,
+> {
+    ////////////////////////////////////////
+    // Interfaces
+    ////////////////////////////////////////
+    runtime: E,
+    consumer: Con,
+    producer: Pro,
+    director: D,
+
+    ////////////////////////////////////////
+    // Outgoing Requests
+    ////////////////////////////////////////
+    /// Mailbox that makes and cancels fetch requests
+    mailbox: mpsc::Receiver<Message<Key>>,
+
+    fetcher: Fetcher<E, C, Key, NetS>,
+
+    ////////////////////////////////////////
+    // Incoming Requests
+    ////////////////////////////////////////
+    serves: FuturesPool<(C::PublicKey, u64, Result<Value, oneshot::Canceled>)>,
+    serve_concurrent: usize,
+
+    // Rate limit for incoming requests per peer
+    #[allow(clippy::type_complexity)]
+    rate_limiter:
+        RateLimiter<C::PublicKey, HashMapStateStore<C::PublicKey>, E, NoOpMiddleware<E::Instant>>,
+
+    ////////////////////////////////////////
+    // Phantom Data
+    ////////////////////////////////////////
+    _s: PhantomData<NetS>,
+    _r: PhantomData<NetR>,
+}
+
+impl<
+        E: Clock + GClock + Spawner + Rng,
+        C: Scheme,
+        D: Director<PublicKey = C::PublicKey>,
+        Key: Array,
+        Con: Consumer<Key = Key, Value = Value, FailureCode = ()>,
+        Pro: Producer<Key = Key, Value = Value>,
+        NetS: Sender<PublicKey = C::PublicKey>,
+        NetR: Receiver<PublicKey = C::PublicKey>,
+    > Actor<E, C, D, Key, Con, Pro, NetS, NetR>
+{
+    pub async fn new(runtime: E, cfg: Config<C, D, Key, Con, Pro>) -> (Self, Mailbox<Key>) {
+        let (sender, receiver) = mpsc::channel(cfg.mailbox_size);
+        let requester = Requester::new(runtime.clone(), cfg.requester_config);
+        let rate_limiter = RateLimiter::hashmap_with_clock(cfg.rate_limit, &runtime);
+        let fetcher = Fetcher::new(
+            runtime.clone(),
+            requester,
+            cfg.fetch_max_outstanding,
+            cfg.fetch_retry_timeout,
+        );
+        (
+            Self {
+                runtime,
+                consumer: cfg.consumer,
+                producer: cfg.producer,
+                director: cfg.director,
+                mailbox: receiver,
+                fetcher,
+                serves: FuturesPool::new(),
+                serve_concurrent: cfg.serve_concurrent,
+                rate_limiter,
+
+                _s: PhantomData,
+                _r: PhantomData,
+            },
+            Mailbox::new(sender),
+        )
+    }
+
+    pub async fn run(&mut self, network: (NetS, NetR)) {
+        let (mut sender, mut receiver) = network;
+        let mut shutdown = self.runtime.stopped();
+
+        loop {
+            // Update peer list
+            self.fetcher.reconcile(&self.director.peers());
```

Comment:
Will be interesting to eventually benchmark this operation (seems like it could get annoyingly expensive on large validator sets to do each loop).

Reply:
(see other comment about returning a reference from `peers()`)